### ISSUE TITLE
Shard a PostgreSQL table in Clickhouse

### ISSUE THREAD

**Issue by MurikiGY** [ClickHouse/ClickHouse#60502]
Hello,

In the Research Center I'm working on we have a PostgreSQL production environment for Foreign Key constraint and we need a DBMS for analytical purposes. Until now I tested the MaterializedPostgreSQL with our databases and everything seems great, but I also wanted to create a cluster for scaling.

My final objective is to have all the tables in PostgreSQL for transactional procedures and a clickhouse cluster sharding these tables for distributed queries.

After setting up a 3 shard 1 replica cluster with clickhouse-keeper I tried to copy the database with the `MaterializedPostgreSQL ON CLUSTER 'cluster_3S_1R'` but instead of sharding the tables I got a full replica on each cluster node, witch is a problem because the data consistency is already on postgres and is using too much disk.

I also tried to use the 
`CREATE TABLE pg_test.pg_table (...) engine = ExternalDistributed('PostgreSQL', ...)` 
and then creating a table using 
`CREATE TABLE pg_test.ch_table AS pg_test.pg_table ENGINE = Distributed(cluster_3S_1R, pg_test, pg_table, rand());`
but when I try to run 
`INSERT INTO pg_test.ch_table SELECT * FROM pg_test.pg_table` 
the client just freeze and when I try to access the ch_table I got the error that there is no table pg_table in the other nodes.
I think I did something wrong but didn't find any other documentation.

In short: How do I shard a Postgres table similar a logical replication system?

Links I used:
1 - https://clickhouse.com/docs/en/guides/sre/keeper/clickhouse-keeper
2 - https://clickhouse.com/docs/en/architecture/cluster-deployment
3 - https://clickhouse.com/docs/en/engines/table-engines/integrations/ExternalDistributed

This is my first time working with clusters so sorry for any trouble or lack of knowledge.

**Comment by UnamedRus** [comment]
>  CREATE TABLE pg_test.ch_table AS pg_test.pg_table ENGINE = Distributed(cluster_3S_1R, pg_test, pg_table, rand());

Distributed table doesn't store data, it's just proxy which allow you to query multiple nodes at once.
For data you need to create MergeTree tables

>  the client just freeze and when I try to access the ch_table I got the error that there is no table pg_table in the other nodes.

All nodes in clickhouse are independent, you need to manually create MergeTree table on them (or use ON CLUSTER statement)


https://youtu.be/Vuh6NOluIxo

**Comment by MurikiGY** [comment]
I see, I though the `Distributed` engine worked with another engines such as `ExternalDistributed`.

I created a MergeTree table, then a Distributed table and inserted with 
`INSERT INTO table SELECT * FROM postgresql('host:port', 'database', 'table', 'user', 'password')`
and it worked just like I wanted, thanks.

But still, when new data is inserted in postgres I'm gonna need to manually insert the data in the distributed table. Is there any setup to automatize this process just like MaterializedPostgres replication?

**Comment by UnamedRus** [comment]
> Is there any setup to automatize this process just like MaterializedPostgres replication?

You can look into https://github.com/Altinity/clickhouse-sink-connector
But it also kind of lacking native Distributed support for all of it's components https://github.com/Altinity/clickhouse-sink-connector/issues/437

**Comment by MurikiGY** [comment]
So it's still in development. Thanks for your help @UnamedRus, I'll try to keep up to date with the Altinity Clickhouse Sink Connector.
